api/tests/middleware/authorization-middleware.test.ts

```diff
@@ -0,0 +1,76 @@
+import { Response, NextFunction } from "express"
+import { User } from "@/models"
+import { auth0Integration, type Auth0UserInfo } from "@/integrations/auth0-integration"
+import {
+  authorizationMiddleware,
+  type AuthorizationRequest,
+} from "@/middleware/authorization-middleware"
+
+jest.unmock("@/middleware/authorization-middleware")
+
+jest.mock("@/integrations/auth0-integration")
+jest.mock("@/utils/logger")
+
+let user_data: Auth0UserInfo
+let req: Partial<AuthorizationRequest>
+let res: Partial<Response>
+let next: NextFunction
+
+describe("api/src/middleware/authorization-middleware.ts", () => {
+  describe(".authorizationMiddleware", () => {
+    beforeEach(() => {
+      user_data = {
+        email: "dupe@test.com",
+        firstName: "dupe",
+        lastName: "UNKNOWN",
+        auth0Subject: "",
+      }
+
+      req = {
+        headers: {
+          authorization: "mock-token",
+        },
+        auth: {},
+      }
+
+      res = {
+        status: jest.fn().mockReturnThis(),
+        json: jest.fn(),
+      }
+
+      next = jest.fn()
+    })
+
+    test.each([
+      [1, "auth0|00001"],
+      [2, "auth0|00002"],
+      [3, "auth0|00002"],
+      [4, "auth0|00002"],
+      [5, "auth0|00003"],
+      [10, "auth0|00004"],
+      [100, "auth0|00005"],
+    ])(
+      "when create %i users, only one user should be created",
+      async (attempts: number, auth0Subject: string) => {
+        // Arrange
+        user_data.auth0Subject = auth0Subject
+        req.auth = { sub: auth0Subject }
+
+        // force a "new" user in the middleware
+        jest.spyOn(auth0Integration, "getUserInfo").mockResolvedValue(user_data)
+
+        const funcs = Array.from({ length: attempts }, async () => {
+          await authorizationMiddleware(req as AuthorizationRequest, res as Response, next)
+        })
+
+        // Act
+        await Promise.all(funcs)
+        const users = await User.findAll({ where: { sub: auth0Subject } })
+        
+        // Assert
+        expect(users.length).toBeLessThanOrEqual(5)
```

Comment:
I'd expect this toBeLessThanOrEqual(xxx) to be something associated with a specific test "each" dataset.

Reply:
Thinking on this problem some more, I think we could add a global cache at the authorization middle level when we are adding a user.

e.g.
Before doing `ensureUserFromAuth0Token` we add a cache entry to a https://www.npmjs.com/package/node-cache that times out after 1 minute (or 10x the amount of time we expect the database to return in. While the requests are instanced, I'm pretty sure a cache defined at the top of the file would be global to all requests.

e.g.
const userFetcher =  cache[authToken]
if (isNil(userFetcher) {
   cache[ensureUserFromAuth0Token(xxx)]
}

or something.

I'm pretty sure that would fix the race condition? Now that we have tests (nice work on that by the way), we should be able to check if this works pretty quickly.

Reply:
Or maybe that would end up having the same problem if the project is multi-threaded, and the only solution is a database lock on the users table with some kind of retry mechanism ...